File: scripts/maze/maze.rb
class Wall

  attr_accessor :value

  def initialize
    @value = true
  end

end

class Cell

  attr_accessor :sides

  def initialize(n_sides)
    @n_sides = n_sides
    @sides = Array.new(@n_sides) { { :wall=>Wall.new, :cell=>nil } }
  end

  def remove_wall(i)
    @sides[i][:wall].value = false
  end

end

class Maze

  attr_accessor :grid
  attr_accessor :solution

  def initialize(rows, cols)
    @rows, @cols = rows, cols
    @grid = Array.new(@cols) { Array.new(@rows) { Cell.new(4) } }
    @grid.each_with_index { |col, ci|
      col.each_with_index { |cell, ri|
        if(ci > 0)
          @grid[ci][ri].sides[1][:cell] = @grid[ci-1][ri]
          @grid[ci][ri].sides[1][:wall] = @grid[ci-1][ri].sides[3][:wall]
        end
        if(ri > 0)
          @grid[ci][ri].sides[0][:cell] = @grid[ci][ri-1]
          @grid[ci][ri].sides[0][:wall] = @grid[ci][ri-1].sides[2][:wall]
        end
        if(ci < @cols-1)
          @grid[ci][ri].sides[3][:cell] = @grid[ci+1][ri]
        end
        if(ri < @rows-1)
          @grid[ci][ri].sides[2][:cell] = @grid[ci][ri+1]
        end
      }
    }
    @solution = Array.new()
  end

  def dfs
    visited = Array.new(@cols) { Array.new(@rows, false) }
    backtrack = Array.new()
    i, j = 0, 0
    until visited[i][j] && backtrack.size < 1
      visited[i][j] = true
      next_cell = [0, 1, 2, 3].collect { |x|
        if (i_new = i + [0, -1, 0, 1][x]).between?(0, @cols-1) &&
             (j_new = j + [-1, 0, 1, 0][x]).between?(0, @rows-1) && !visited[i_new][j_new]
          {:direction=>x, :coordinates=>[i_new,j_new]}
        end
      }.compact.choice
      if next_cell
        backtrack.push([i,j])
        @grid[i][j].remove_wall(next_cell[:direction])
        i, j = next_cell[:coordinates]
        next
      elsif backtrack != []
        if [i, j] == [@cols-1, @rows-1]
            @solution = backtrack.dup.push([i, j])
        end
        i, j = backtrack.pop
        next
      end
    end
  end

  CELL_WIDTH = ARGV[2].to_i
  CELL_HEIGHT = ARGV[3].to_i

  def to_svg
    svg_list = @grid.each_with_index.collect { |col, ci|
      col.each_with_index.collect { |cell, ri|
        [[[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0, 1], [1, 1]], [[1, 1], [1, 0]]].select.with_index { |edge, i|
          (i!=2 || ri==@rows-1) && (i!=3 || ci==@cols-1) && cell.sides[i][:wall].value
        }.collect { |c1, c2|
          "id=\"#{[ci, ri].math_op(c1.math_op(c2, :+).map { |x| x/2 }, :+).to_custom_s}" <<
            ["v", "h"][c1.common(c2)[0]] << "\"" <<
            " x1=\"#{CELL_WIDTH*(ci+c1[0])}\" y1=\"#{CELL_HEIGHT*(ri+c1[1])}\"" <<
            " x2=\"#{CELL_WIDTH*(ci+c2[0])}\" y2=\"#{CELL_HEIGHT*(ri+c2[1])}\""
        }
      }
    }.flatten.unshift(
        @solution.collect { |point|
                "#{CELL_WIDTH*(point[0]+0.5)},#{CELL_HEIGHT*(point[1]+0.5)}"
        }.intersperse(" ").join
    )
    svg_list.intersperse("\n").join

  end

end

class Array

  def intersperse(item)
    self.zip(Array.new(self.size, item)).flatten[0...-1]
  end

  def to_custom_s
    self.collect { |item| item.to_s }.intersperse(",").join
  end

  def math_op(ar, op)
    self.zip(ar).map { |x| x.reduce(op) }
  end

  def common(ar)
    self.zip(ar).each_with_index.collect { |item, index | if item[0]==item[1] then index end }.compact
  end

end

m = Maze.new(ARGV[0].to_i, ARGV[1].to_i)
m.dfs
print m.to_svg

File: scripts/maze/maze.php
<?php

class Wall {

    public $val;

    function __construct() {
        $this->val = true;
    }

}

class Cell {

    public $sides;

    private $n_sides;

    function __construct($n_sides) {
        $this->n_sides = $n_sides;
        $this->sides = array();
        foreach(range(0, $this->n_sides-1) as $i) {
            array_push($this->sides, array("wall"=>new Wall, "cell"=>null));
        }
    }

    function remove_wall($i) {
        $this->sides[$i]["wall"]->val = false;
    }
}

class Maze {

    public $grid, $solution;

    private $rows, $cols, $cell_width, $cell_height;

    function __construct($rows, $cols, $cell_width, $cell_height) {
        $this->rows = $rows;
        $this->cols = $cols;
        $this->cell_width = $cell_width;
        $this->cell_height = $cell_height;

        $this->grid = array();

        foreach(range(0, $this->cols-1) as $i) {
            $col = array();
            foreach(range(0, $this->rows-1) as $j) {
                array_push($col, new Cell(4));
            }
            array_push($this->grid, $col);
        }
        foreach($this->grid as $ci=>$col) {

            foreach($col as $ri=>$row) {

                if($ci > 0) {
                    $this->grid[$ci][$ri]->sides[1]["cell"] = $this->grid[$ci-1][$ri];
                    $this->grid[$ci][$ri]->sides[1]["wall"] = $this->grid[$ci-1][$ri]->sides[3]["wall"];
                }
                if($ri > 0) {
                    $this->grid[$ci][$ri]->sides[0]["cell"] = $this->grid[$ci][$ri-1];
                    $this->grid[$ci][$ri]->sides[0]["wall"] = $this->grid[$ci][$ri-1]->sides[2]["wall"];
                }
                if($ci < $this->cols-1) {
                    $this->grid[$ci][$ri]->sides[3]["cell"] = $this->grid[$ci+1][$ri];
                }
                if($ri < $this->rows-1) {
                    $this->grid[$ci][$ri]->sides[2]["cell"] = $this->grid[$ci][$ri+1];
                }

            }
        }
        $this->solution = array();
        $this->dfs();
    }

    private function dfs() {
        $visited = array();
        foreach(range(0, $this->cols-1) as $cols) {
            $col = array();
            foreach(range(0, $this->rows-1) as $rows) {
                array_push($col, false);
            }
            array_push($visited, $col);
        }
        $backtrack = array();
        $i = 0;
        $j = 0;
        while(!($visited[$i][$j] && sizeof($backtrack)<1)) {

            $visited[$i][$j] = true;
            $coord_x = array(0, -1, 0, 1);
            $coord_y = array(-1, 0, 1, 0);
            $possibles = array();
            foreach(range(0, 3) as $x) {
                $i_new = $i + $coord_x[$x];
                $j_new = $j + $coord_y[$x];
                if($i_new>=0 && $i_new<=$this->cols-1 && $j_new>=0 && $j_new<=$this->rows-1) {
                    if(!$visited[$i_new][$j_new]) {
                        array_push($possibles, array("direction"=>$x, "coordinates"=>array($i_new, $j_new)));
                    }
                }
            }
            if(sizeof($possibles)>0) {
                $next_cell = $possibles[array_rand($possibles)];
                array_push($backtrack, array($i, $j));
                $this->grid[$i][$j]->remove_wall($next_cell["direction"]);
                $i = $next_cell["coordinates"][0];
                $j = $next_cell["coordinates"][1];
                continue;
            }
            elseif(sizeof($backtrack) != 0) {
                if($i == $this->cols-1 && $j == $this->rows-1) {
                    $this->solution = $backtrack;
                    array_push($this->solution, array($i, $j));
                }
                $next_cell = array_pop($backtrack);
                $i = $next_cell[0];
                $j = $next_cell[1];
                continue;
            }
        }
    }

    public function write_to(&$arr) { 
        $polyline = "";
        foreach($this->solution as $point) {
            $polyline .= $this->cell_width*($point[0]+0.5) . "," . $this->cell_height*($point[1]+0.5) . " ";
        }

        array_push($arr, $polyline);

        $id_array = array(array(array(1, 0), array(0, 0)),
                          array(array(0, 0), array(0, 1)),
                          array(array(0, 1), array(1, 1)),
                          array(array(1, 1), array(1, 0)));
        foreach($this->grid as $ci=>$col) {
            foreach($col as $ri=>$cell) {
                $lines = array();
                foreach($id_array as $i=>$edge) {
                    if($i != 2 || $ri==$this->rows-1) {
                        if($i !=3 || $ci==$this->cols-1) {
                            if($cell->sides[$i]["wall"]->val) {
                                array_push($lines, $edge);
                            }
                        }
                    }
                }
                foreach($lines as $edge) {
                    $c1 = $edge[0];
                    $c2 = $edge[1];
                    
                    $cx = $ci+floor(($c1[0]+$c2[0])/2);
                    $cy = $ri+floor(($c1[1]+$c2[1])/2);

                    $dir = ($c1[0] == $c2[0]) ? "v" : "h";

                    $id = $dir . $cx . "_" . $cy;

                    $x1 = $this->cell_width*($ci+$c1[0]);
                    $x2 = $this->cell_width*($ci+$c2[0]);
                    $y1 = $this->cell_height*($ri+$c1[1]);
                    $y2 = $this->cell_height*($ri+$c2[1]);

                    array_push($arr, "id=\"" . $id . "\" x1=\"" . $x1 . "\" x2=\"" . $x2 . "\" y1=\"" . $y1 . "\" y2=\"" . $y2 . "\"");
                }
            }
        }

    }
}
